pkg/cloudprovider/cloudapi/azure/azure_security.go

```diff
@@ -344,100 +336,172 @@ func (computeCfg *computeServiceConfig) buildEffectiveNSGSecurityRulesToApply(ap
 	currentNsgSecurityRules := nsgObj.Properties.SecurityRules
 	appliedToGroupNepheControllerName := appliedToGroupID.GetCloudName(false)
-	azurePluginLogger().Info("Building peering security rules", "applied to security group", appliedToGroupNepheControllerName)
+	azurePluginLogger().Info("Building security rules", "applied to security group", appliedToGroupNepheControllerName)
 	for _, rule := range currentNsgSecurityRules {
 		if rule.Properties == nil || *rule.Properties.Priority == vnetToVnetDenyRulePriority {
 			continue
 		}
 
-		// Nephe rules will be created from ruleStartPriority and have description.
+		// check if the rule is Nephe rules based on ruleStartPriority and description.
 		desc, ok := securitygroup.ExtractCloudDescription(rule.Properties.Description)
 		if *rule.Properties.Priority >= ruleStartPriority && ok {
-			// skip any rules created by current processing appliedToGroup (as we have new rules for this group).
-			ruleAddrGroupName := desc.AppliedToGroup
+			// check if the rule is created by current processing appliedToGroup.
 			_, _, isNepheControllerCreatedRule := securitygroup.IsNepheControllerCreatedSG(desc.AppliedToGroup)
-			if isNepheControllerCreatedRule && strings.Compare(ruleAddrGroupName, appliedToGroupNepheControllerName) == 0 {
-				continue
+			if isNepheControllerCreatedRule && strings.Compare(desc.AppliedToGroup, appliedToGroupNepheControllerName) == 0 {
+				// skip the rule if found in remove list.
+				removeRules := rmIngressSecurityRules
+				if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionOutbound {
+					removeRules = rmEgressSecurityRules
+				}
+				idx, found := findSecurityRule(rule, removeRules)
+				if found {
+					// remove the rule from remove list to avoid redundant comparison.
+					removeRules[idx] = nil
+					continue
+				}
 			}
 		}
 
-		// preserve rules.
+		// preserve current rules that do not need to update.
 		if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionInbound {
-			currentNsgIngressRules = append(currentNsgIngressRules, *rule)
+			currentNsgIngressRules = append(currentNsgIngressRules, rule)
 		} else {
-			currentNsgEgressRules = append(currentNsgEgressRules, *rule)
+			currentNsgEgressRules = append(currentNsgEgressRules, rule)
 		}
 	}
 
-	agAsgMapByNepheControllerName, _, err := getNepheControllerCreatedAsgByNameForResourceGroup(computeCfg.asgAPIClient, rgName)
+	allIngressRules := updateSecurityRuleNameAndPriority(currentNsgIngressRules, addIngressSecurityRules)
+	allEgressRules := updateSecurityRuleNameAndPriority(currentNsgEgressRules, addEgressSecurityRules)
+
+	return append(allIngressRules, allEgressRules...), nil
+}
+
+// buildEffectivePeerNSGSecurityRulesToApply prepares the update rule cloud api payload from internal rules that require peering.
+func (computeCfg *computeServiceConfig) buildEffectivePeerNSGSecurityRulesToApply(appliedToGroupID *securitygroup.CloudResourceID,
```

Comment:
We dont exercise/support peering code right, for non peer case

Reply:
correct. We don't claim support for peering, its just that the peering code is here and I just update them too.